### ISSUE TITLE
[@typescript-eslint/no-unused-vars] Option to allow phantom type parameters

### ISSUE THREAD

**Issue by calebmer** [typescript-eslint/typescript-eslint#2634]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

An example class for declaring a `Type` with phantom `Context` and `Value` parameters. For a hypothetical GraphQL-like framework in Node.js.

```JSON
{
  "rules": {
    "@typescript-eslint/no-unused-vars": "warn"
  }
}
```

```TS
abstract class Type<Context, Value> {
  public abstract isOutput(): boolean;
  public abstract isInput(): boolean;
}
```

**Expected Result**

This is a feature request, I’d like an option to ignore phantom type parameters on classes and type aliases like `Context` and `Value` in this case. I do still want the rule to warn for unused type parameters on function declarations.

**Actual Result**

TypeScript reports an unused variable warning.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.1.1` |
| `@typescript-eslint/parser`        | `4.1.1` |
| `TypeScript`                       | `4.0.2` |
| `ESLint`                           | `7.9.0` |
| `node`                             | `14.8.0` |

**Comment by bradzacher** [comment]
You can use `varsIgnorePattern` to define an ignore pattern (like `^_`) and name the variables appropriately.

https://eslint.org/docs/rules/no-unused-vars#varsignorepattern

We won't be adding an option to specifically ignore unused type parameters.

**Comment by jawj** [comment]
The thing about unused type parameters, though, is that they're not necessarily unused. For example, in one of my projects, this gets flagged:

```typescript
export interface TxnClient<T extends Isolation> extends pg.PoolClient { }
```

But the type parameter is in fact well used by functions that pass `TxnClient` instances around, for example:

```typescript
export async function transaction<T, M extends Isolation>(
  pool: pg.Pool,
  isolationMode: M,
  callback: (client: TxnClient<M>) => Promise<T>
): Promise<T> {
  /* ... */
}
```

What do you suggest is the answer in this case? Using an ignore pattern seems like a nasty hack when there's nothing I can see wrong with the usage.

**Comment by bradzacher** [comment]
Let me frame it another way for you.

```ts
function foo(a: string, b: number) {
  return a;
}

foo('a', 1);
```

In the above code, is the parameter `b` unused?

By your definition it's not unused because one of the callsites passes the argument in - but that is very clearly wrong.

---

The definition of unused that the rule uses is simple "a variable that is defined and has no references in a child scope".

For the type
```
interface Foo<T> {}

type Usage = Foo<string>;
```
`T` is defined, but is never used in a child scope. The fact that a "callsite" passes the parameter in does not change that fact.

-----

If you don't like configuring the rule with the standard underscore naming pattern, then you can use another pattern - the option simply accepts a regex.

For example, you could make it so you have to explicitly name them as phantom:
`varsIgnorePattern: "Phantom$"`
This would be less "hacky" and serve as very clear documentation that the variable has 100% been intentionally written as a phantom parameter, and not just forgotten about.
 
```
interface Foo<TPhantom> {}
```

**Comment by jawj** [comment]
Thanks for the detailed response.

I can see that your sensible definition of 'unused' covers this situation — but I'm not sure the analogy with function calls is quite right. 

After all, in the type variable case, when the callsite passes the parameter, it gets back something that depends on that parameter it passed, regardless of whether or not the parameter was used within the callee. The returned type is effectively tagged with the type that was passed in. For this reason, I think it would be helpful if linting this situation could be handled separately. 

At the end of the day, though, you're the maintainer, and you've got to make the call. Thanks again.
